### PR TITLE
RFC: spaceTx glossary

### PR DESCRIPTION
**Pull request by ambrosejcarr**
This is an early PR to unblock collaborators @joshmoore @hewgreen @zperova

We are open to changing terminology that users find confusing, or to better fit the existing ecosystem. Please comment!

### CODE REVIEW

File: docs/glossary.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+a document describing the above terms in detail, i.e., this document.
+
+### Pipeline Implementation
+actual code for the pipeline. This code will be packaged as a well-documented Python library and corresponding command line tool for use by consortium members to facilitate easy sharing and comparison of results across labs/methods.
+
+### Manifest
+The data manifest is a file that includes necessary information about the hybridization images, auxiliary images, and codebook. The SpaceTx manifest format is described in detail abovebelow.
+
+### Image Tile
+a single plane, single channel 2D image. In the manifest, each tile has information about it’s (X,Y,Z) coordinates in space, and information about which hybridization round (H) and/or fluorescence channel (C) it was acquired under.
```
Comment (berl) [inline docs/glossary.md]: I think we prefer `R` over `H` now, right?

**Comment by ttung** [comment]
There are still a bunch of mentions of hybridization images.  I think primary images might be an apt term, but I suck at naming things.

File: docs/glossary.md (hunk 1 above)
Comment (ambrosejcarr) [reply docs/glossary.md]: Yep, sorry. I just copied and pasted this in early in the AM. I will amend.

File: docs/glossary.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+### Pipeline Specification
+a document describing the above terms in detail, i.e., this document.
+
+### Pipeline Implementation
+actual code for the pipeline. This code will be packaged as a well-documented Python library and corresponding command line tool for use by consortium members to facilitate easy sharing and comparison of results across labs/methods.
+
+### Manifest
+The data manifest is a file that includes the locations of all fields of view for either primary or auxiliary images. 
+
+### Image Tile
```
Comment (ambrosejcarr) [inline docs/glossary.md]: @hewgreen this is the Tile description.

Comment (hewgreen) [reply docs/glossary.md]: thanks

**Comment by joshmoore** [comment]
Barring your alphabetization, my primary concern would be use of `tile` which for me/us is a subregion of a plane (your tile).

**Comment by ambrosejcarr** [comment]
@joshmoore How would you rename it? I don't think it appears very frequently in our codebase. If people will be confused I'm open to modifications. 

(If you have other smaller concerns please list those as well!)

**Comment by hewgreen** [comment]
This is useful. There were a few surprises for me:

Tiles are 3D, sort of- good to know! Is the depth readily available in pixels or uM?

FOV is not the largest entity (because they can be stitched)- I presumed the spacetx pipeline did the stitching. So in the README when I read a FOV is 'the most complex file in the spaceTx format' I presumed they were stitched which confused the term for me. My mistake.

Target- is more than target type (mRNA or protein). I'd like you to define this more rigorously if you can. If I get a target like 'SCUBE2' can I use this to find the transcript or probe sequence? I think here the HCA will require either a general sub cellular structure (e.g. like DNA) or a more specific ID (protein or transcript). We will also likely need probe information.

**Comment by ambrosejcarr** [comment]
@hewgreen 

> Tiles are 3D, sort of- good to know! Is the depth readily available in pixels or uM?

I think we're on the same page here, but it's a 2d array, to be crystal clear. The depth is so-far found in the HCA metadata, and not in starfish's internal metadata. 

> I presumed the spacetx pipeline did the stitching

Yea, we've punted on this for now. Some day that will hopefully be true. 

>  Target- is more than target type (mRNA or protein). I'd like you to define this more rigorously if you can. If I get a target like 'SCUBE2' can I use this to find the transcript or probe sequence? I think here the HCA will require either a general sub cellular structure (e.g. like DNA) or a more specific ID (protein or transcript). We will also likely need probe information.

Actually, we'd love some help here. I think you guys have a better sense of what ontologies are available to describe these things. I can put up a straw man for a `probes.sam` file which links to target in some way, but that's very much still an open question. Do you and @zperova have time to help me solve this?

File: docs/glossary.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+a single plane, single channel 2D image. In the manifest, each tile has information about it’s (X,Y,Z) coordinates in space, and information about which imaging round (R) and/or fluorescence channel (C) it was acquired under.
+
+### Field of View (FOV)
+a collection of Image Tiles corresponding to a specific volume or plane of the sample, under which the signal for all channels and all imaging rounds were acquired. All tiles within this FOV are the same size, but the manifest allows for different spatial coordinates for different imaging rounds or channels (to accommodate slight movement between rounds, for example).
+In microscopy, a field of view corresponds to the area of the camera sensor, and many such fields of view are expected to be taken per tissue slice. 
+
+### Coordinates (field of view):
```
Comment (berl) [inline docs/glossary.md]: The coordinates are actually per tile, which is correct in the Image Tile description, but a little confusing here.

**Comment by joshmoore** [comment]
> @joshmoore How would you rename it?

On our side:
 - XSD: "[Plane](https://www.openmicroscopy.org/Schemas/Documentation/Generated/OME-2016-06/ome_xsd.html#Plane)"
 - Docs: "[focal plane"](https://docs.openmicroscopy.org/ome-model/5.6.3/ome-xml/index.html)
 - Original paper: "[optical plane](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2005-6-5-r47#Fig1)"

but I'm _completely_ happy with a thesaurus strategy: naming is a hard problem.

**Comment by zperova** [comment]
@ambrosejcarr Could you please point us to the probes.sam file?

**Comment by ambrosejcarr** [comment]
@dganguli @berl @ttung Any complaints about changing our terminology from "Tile" -> "Plane"?

**Comment by ambrosejcarr** [comment]
@zperova @hewgreen Right now probes.sam only lives here: https://docs.google.com/document/d/12V-BQF-wh1GDBXwUH2hQrPhw6n0oQqENPK-YHGv59w4/edit#bookmark=id.xmtuxekc7a1c

**Comment by berl** [comment]
I prefer "Tile". "Plane" referring to a rectangular polygon in the x-y *plane* is confusing to me.  If this is the only bit that is muddled with the OME metadata and it can be addressed with a simple thesaurus as @joshmoore  says, then I think we're in pretty good shape.

File: docs/glossary.md (hunk 3 above)
Comment (ambrosejcarr) [reply docs/glossary.md]: Ah, so the intuition would be that the coordinates should be able to capture jitter across Tiles in the field of view?

File: docs/glossary.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+a sequence of data processing steps to process the inputs into the desired outputs.
+
+### Pipeline Component
+a single data processing step in the pipeline, as defined by its input and output file formats, e.g., the spot-detection component takes as input an image and outputs a table of spot locations, shapes, and intensities.
+
+### Pipeline Component Algorithm
+a specific algorithm type that adheres to the specified input and output file formats required by the component it belongs to. For example, a spot-detection component algorithm can be realized as a Gaussian blob detector or a connected components labeller. Both find spots and accept the same inputs and produce the same outputs, hence belong to the same component. However, the underlying properties of the algorithms (and parameterizations) may be quite different.
+
+### Pipeline Specification
+a document describing the above terms in detail, i.e., this document.
```
Comment (dganguli) [inline docs/glossary.md:33]: a document describing the pipeline in detail, including an ordered list of all the pipeline components, and expected input/output file formats at each step of computation.

File: docs/glossary.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
+## Glossary
+
+Here, we clearly define the relevant terms needed to understand the preceding pipeline specification (in no particular order).
```
Comment (dganguli) [inline docs/glossary.md:3]: "... to understand the spaceTx pipeline specification" ?

File: docs/glossary.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
+## Glossary
```
Comment (dganguli) [inline docs/glossary.md:1]: missing definitions of: 
### feature 
the value of a spot (aggregated across all pixel values circumscribed by that spot) or the value of a single pixel.

### putative barcode (or some other name for this)
feature values across all imaging rounds and/or color channels. these map to codewords in a codebook.

### decoding
matching putative barcodes to codewords in a codebook to read out the corresponding target believed to be associated with that barcode.

Comment (dganguli) [inline docs/glossary.md:1]: glossary could be sub-sectioned by:

- stuff about pipelines

- stuff about tiles, fovs, rounds, channels

- stuff about features, barcodes, codebooks, decoding

order probably matters to bulid a logical narrative flow.

Comment (ambrosejcarr) [reply docs/glossary.md:1]: > decoding
matching putative barcodes to codewords in a codebook to read out the corresponding target believed to be associated with that barcode.

Ok, this is a good idea. Should we describe each of the different API stages as we envision them? (e.g. image: filter, segment, register, spots: detect, decode, assign) Or perhaps that goes into starfish?

Comment (ambrosejcarr) [reply docs/glossary.md:1]: Do you think stuff about pipelines belongs here? or just a minimum to describe the objects of the spec, and more detail goes in a starfish glossary?

**Comment by dganguli** [comment]
@joshmoore that paper you linked to is very nice -- thanks! i'm embarrassed to say i hadn't seen it yet. i'll study it more carefully. we have a lot to learn from you :)

@ambrosejcarr @berl i have no strong opinion about tile vs. plane, other than tile appears to be the path of least resistance.

File: docs/glossary.md (hunk 6 above)
Comment (dganguli) [reply docs/glossary.md:1]: yes, but i think that belongs elsewhere. probably in some actual pipeline specification document.

Comment (dganguli) [reply docs/glossary.md:1]: i think so. then some pipeline spec doc and/or starfish thing can reference this one global glossary. we can explicitly link to sub-sections of it as well maybe.

**Review by dganguli — APPROVED** [review]
awesome. thanks @ambrosejcarr ! left some comments but nothing high priority.

**Comment by joshmoore** [comment]
> that paper you linked to is very nice -- thanks!

No problem, @dganguli. It's obviously a bit long in the tooth, but it will still inform a large portion of our software stack.

**Comment by ambrosejcarr** [comment]
As identified by @dganguli, this document requires additional work. I am incorporating it in its current form (but also including responses to comments) into the docs in the branch https://github.com/spacetx/starfish/tree/ajc-readthedocs

**Pull request closed without merging**